src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ReflectMethods.java

```diff
@@ -525,16 +529,113 @@ class BodyScanner extends FilterScanner {
             this.name = names.fromString("quotedLambda");
             this.isQuoted = true;
 
-            com.sun.tools.javac.util.List<Type> nil = com.sun.tools.javac.util.List.nil();
-            MethodType mtype = new MethodType(nil, syms.quotedType, nil, syms.methodClass);
+            QuotableLambdaCaptureScanner lambdaCaptureScanner =
+                    new QuotableLambdaCaptureScanner(tree);
+
+            List<VarSymbol> capturedSymbols = lambdaCaptureScanner.analyzeCaptures();
+            int blockArgOffset = 0;
+
+            ListBuffer<Type> capturedTypes = new ListBuffer<>();
+            if (lambdaCaptureScanner.capturesThis) {
+                capturedTypes.add(currentClassSym.type);
+                blockArgOffset++;
+            }
+            for (Symbol s : capturedSymbols) {
+                capturedTypes.add(s.type);
+            }
+
+            MethodType mtype = new MethodType(capturedTypes.toList(), syms.quotedType,
+                    com.sun.tools.javac.util.List.nil(), syms.methodClass);
             FunctionType mtDesc = FunctionType.functionType(typeToTypeElement(mtype.restype),
                     mtype.getParameterTypes().map(this::typeToTypeElement));
 
             this.stack = this.top = new BodyStack(null, tree.body, mtDesc);
 
+            // add "this" capture (if needed)
+            if (lambdaCaptureScanner.capturesThis) {
+                top.localToOp.put(currentClassSym, null); // @@@ just make translation happy
```

Comment:
I think we can avoid this. For a quotable lambda we know the number of block parameters (captures) before we obtain the captured expressions from `top.localToOp`.

If `funcOp.parameters().size() == bodyScanner.stack.localToOp.size() + 1` then we know we need to first capture `this` expression. This is easier to do if we inline the call to `quotedCapturedArgs`.

Reply:
That said - not sure. Isn't `funcOp` the function for the lambda? That should not feature any captured argument - just the parameter of the lambda as they appear in the source. I'm not sure we can deduce the need for `this` capture just by looking at that?

Reply:
I think my comment was confusing. Here's another way of describing it - you could add the following to `quotedCapturedArgs` 
```
        assert bodyScanner.top.body.entryBlock().parameters().size() == bodyScanner.top.localToOp.keySet().size();
```
The top body becomes the body of the `quotedLambda` func op.

Reply:
So i think we could avoid `top.localToOp.put(currentClassSym, null)` and detect if we need to synthesize `this` by checking the sizes.

Reply:
Ah yes, constants, i see you count afterwards.